Improve task library grid [SCHOOL-162]

BillClinton opened the pull request:
Make improvements to the Task Manager Grid including: 
 * Configure Sencha’s built-in [grid filter plugin](https://docs.sencha.com/extjs/6.2.0/classic/Ext.grid.filters.Filters.html)

 * Fix joining of skills into comma-delimited list

 * Implement reasonable sorter functions for most/all columns

 * Make sure paging works

 * Implement a resizable split view between grid and right-side details panel

BillClinton committed:
wip: ui adjustments

BillClinton committed:
fix: remove header added in previous commit

BillClinton committed:
feat: show column seperators in header

This makes it easier to match column data with the header lable

BillClinton committed:
feat: adjust alignment and widths

BillClinton committed:
fix: align detail header with grid headers

BillClinton committed:
fix: align detail header with grid headers

This is a different approach to the change made in ebf5ea13bc450633a4d620e23fcacb9edcb2b367 .
This approach better reflects the manner in which padding/margins is handled
in the Ext JS grid headers.

BillClinton committed:
fix: update app title class name

BillClinton committed:
wip: temp fix for header padding

BillClinton committed:
fix: add pageSize param to activate paging

BillClinton committed:
fix: remove default flex overriding fixed widths

BillClinton committed:
wip: add hack to set vertical size of grid

This needs to be removed and configured so the grid fits the size
of the user's screen vertically

BillClinton committed:
feat: add tooltip for skills column

BillClinton committed:
wip: create fixed width viewport for testing

BillClinton committed:
feat: restructure app layout

Add an outer container in order to position header, grid and details panes
in a border layout, allowing the details pane to be resizeable.

BillClinton committed:
feat: add column renderer to list skills

BillClinton committed:
feat: add filtering to task grid component

BillClinton committed:
feat: allow task searching by parent task

BillClinton committed:
feat: add search filter for skills column

BillClinton committed:
feat: add toolbar button to clear grid filters

BillClinton committed:
wip: add non functioning date filter w/ todo note

BillClinton committed:
feat: display total results in bottom toolbar

BillClinton committed:
fix: prevent button from moving results display

BillClinton committed:
feat: enable/disable buttons based on selection

BillClinton committed:
feat: add filter for task creator

BillClinton committed:
fix: correct query and account for zero results

BillClinton committed:
fix: remove Saki SCSS for retired search component

BillClinton committed:
feat: add sorting by parent task

BillClinton committed:
feat: add sorting by task creator

BillClinton committed:
feat: add sorting by experience type

BillClinton committed:
feat: add sorting by skills

BillClinton committed:
refactor: clean up Creator sorter function

BillClinton committed:
refactor: clean up skills sorter function

BillClinton committed:
refactor: clean up experience type sorter function

BillClinton committed:
refactor: clean up parent task sorter function

BillClinton committed:
fix: correct id issue when multiple active filters

BillClinton committed:
fix: remove collapsible config

This is a split panel which can be resized.  I found that also making it
collapsible was kind of confusing and provided more flash than utility

BillClinton committed:
feat: add temp css for proper grid sizing

BillClinton committed:
fix: correct ambiguous field name error

Filtering by Title would cause an ambiguous field error when multiple columns were filtered.
I Had to move this query to a callback so I could insert the table alias in the query.

BillClinton committed:
refactor: remove superfluous param from sorters

BillClinton committed:
refactor: remove superfluous param from filters

BillClinton committed:
fix: update test with new class name

The name of the app header title class was updated/corrected recently so it
would use the same class as the app header titles in other parts of Slate.

BillClinton committed:
feat: remove temp inline css from index.html

BillClinton committed:
feat: add styling for border layout viewport fit

BillClinton committed:
feat: allow inclusion of archived tasks

BillClinton committed:
feat: add checkbox to include archived tasks

BillClinton committed:
feat: update param when archived checkbox clicked

BillClinton committed:
feat: add param to store for filtering by archived

BillClinton committed:
refactor: clean up tasks controller

BillClinton committed:
fix: correct indent

BillClinton committed:
feat: move archive checkbox to a menubutton

BillClinton committed:
feat: restore app header after moving checkbox

BillClinton committed:
fix: remove unneeded scss

BillClinton committed:
fix: correct splitter appearing above slate menu

BillClinton committed:
refactor: remove unused componentCls config

BillClinton committed:
fix: correct typo in css class name

BillClinton committed:
feat: implement archive event handlers

BillClinton committed:
fix: correct indentation

BillClinton committed:
feat: allow admin user to browse archived tasks

BillClinton committed:
fix: update component reference in test

BillClinton committed:
note: Add todo pointing out race condition issues

BillClinton reviewed (commented):
I made a note about the two race conditions that are causing test failures.  We can be pretty sure that these are race conditions because if I duplicate the previous focus command so it is performed twice, there are no test failures. @themightychris

BillClinton committed:
fix: correct race condition in task manager test

BillClinton committed:
fix: correct clipping of reload icon

themightychris commented on php-classes/Slate/CBL/Tasks/Task.php:
```diff
@@ -233,4 +262,135 @@ protected static function validateTaskStatus(RecordValidator $RecordValidator, T
+    {
+        $parentTasks = DB::allRecords('SELECT ID FROM %s WHERE Title LIKE "%%%s%%"', [
+            Task::$tableName,
+            $identifier
```
```suggestion
            DB::escape($identifier)
```

themightychris commented on php-classes/Slate/CBL/Tasks/Task.php:
```diff
@@ -233,4 +262,135 @@ protected static function validateTaskStatus(RecordValidator $RecordValidator, T
+        [
+            TaskSkill::$tableName,
+            Skill::$tableName,
+            $identifier
```
```suggestion
            DB::escape($identifier)
```

themightychris commented on php-classes/Slate/CBL/Tasks/Task.php:
```diff
@@ -233,4 +262,135 @@ protected static function validateTaskStatus(RecordValidator $RecordValidator, T
+        [
+            Task::$tableName,
+            Person::$tableName,
+            $identifier,
+            $identifier
```
```suggestion
            DB::escape($identifier),
            DB::escape($identifier)
```

themightychris reviewed (changes requested):
@BillClinton looks great, fabulous work!

The only major issue I see is that we're putting user input directly into filter SQL without escaping, that will need fixing before merge

themightychris committed:
chore: delete unused scss file

themightychris committed:
refactor: move scss next to JS

themightychris committed:
fix: get rid of extra autoCreate

themightychris committed:
fix: replicate SlateAdmin pattern for full-screen app with omnibar

themightychris committed:
refactor: remove z-index fix to splitter

SlateAdmin viewport code comes with an adjustment to the omnibar z-index to more broadly play nice with default Sencha z-indexes

themightychris commented on sencha-workspace/SlateTasksManager/sass/src/view/TasksManager.scss:
```diff
@@ -1,9 +1,23 @@
+  bottom: 0px;
+  width: 100vw;
+}
+
```
in stuff within `app/sass/{src,var}`, each file corresponds to a component and it's best to avoid including any CSS anywhere in this tree that wouldn't be scoped to the component the file is for. Notice the version before this only had stuff within `.slate-tasks-manager`

`app/sass/etc/all.scss` is the place for random global stuff

In any case, I found this calculation to be off a bit and leaving gaps in the window, so I've replicated how SlateAdmin over in the `slate` repository is set up and did some other clean up: https://github.com/SlateFoundation/slate-cbl/pull/684/files/b6b42921ddcf39a45ec8d20e35fa007c7091f860..e7544db3d673dbfa063bac38ad6d8d3c65242a21

themightychris committed:
fix(e2e): update test selectors
